src/protocol/ipa/mod.rs

```diff
@@ -771,132 +771,195 @@ pub mod tests {
+                Some(uid) => assert_eq!(uid, record.user_id, "All records passed to `update_expected_output_for_user` should share a common `user_id`."),
+                None => last_user_id = Some(record.user_id),
+            }
 
-            for _ in 0..NUM_USERS {
-                let random_user_id = rng.gen_range(0..MAX_USER_ID);
-                let num_records_for_user = rng.gen_range(1..MAX_RECORDS_PER_USER);
-                let mut records_for_user = Vec::with_capacity(num_records_for_user);
-                for _ in 0..num_records_for_user {
-                    let random_timestamp = rng.gen_range(0..SECONDS_IN_EPOCH);
-                    let is_trigger_report = rng.gen::<bool>();
-                    let random_breakdown_key = if is_trigger_report {
-                        0
-                    } else {
-                        rng.gen_range(0..MAX_BREAKDOWN_KEY)
-                    };
-                    let trigger_value = if is_trigger_report {
-                        rng.gen_range(1..MAX_TRIGGER_VALUE)
-                    } else {
-                        0
-                    };
-                    records_for_user.push(TestRawDataRecord {
-                        user_id: random_user_id,
-                        timestamp: random_timestamp,
-                        is_trigger_report,
-                        breakdown_key: random_breakdown_key,
-                        trigger_value,
-                    });
+            match last_timestamp {
+                Some(ts) => {
+                    assert!(record.timestamp <= ts, "Records passed to `update_expected_output_for_user` should be in reverse chronological order");
+                    last_timestamp = Some(record.timestamp);
                 }
+                None => last_timestamp = Some(record.timestamp),
+            }
 
-                // sort in reverse time order
-                records_for_user.sort_unstable_by(|a, b| b.timestamp.cmp(&a.timestamp));
-
-                let mut pending_trigger_value = 0;
-                let mut total_contribution = 0;
-                for record in &records_for_user {
-                    if total_contribution >= per_user_cap {
-                        continue;
-                    }
-
-                    if record.is_trigger_report {
-                        pending_trigger_value += record.trigger_value;
-                    } else if pending_trigger_value > 0 {
-                        let delta_to_per_user_cap = per_user_cap - total_contribution;
-                        let capped_contribution =
-                            std::cmp::min(delta_to_per_user_cap, pending_trigger_value);
-                        expected_results[record.breakdown_key] += capped_contribution;
-                        total_contribution += capped_contribution;
-                        pending_trigger_value = 0;
-                    }
-                }
+            if total_contribution >= per_user_cap {
+                continue;
```

Comment:
```suggestion
                break;
```

Reply:
Oh nice. Now that we've refactored this into a separate function I can easily do that. It was a bit gnarly when it was inside a doubly nested for loop before =)